Bug 1863011: Fix checking if ClusterRoleBindings need update

paulfantom opened the pull request:
<!--
    Don't forget about CHANGELOG if this affects the end user!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Monitoring <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR
    <Component> Component affected by your changes such as deps bump, alerts changes and any user facing changes.

    Example:
    - [#741](https://github.com/openshift/cluster-monitoring-operator/pull/741) Bump thanos components to v0.11.0 release
-->

* [ ] I added CHANGELOG entry for this change.
* [x] No user facing changes, so no entry in CHANGELOG was needed.

Fixing CRB re-creation checks by using the same logic as we have for RoleBindings.

s-urbaniak commented on pkg/client/client.go:
```diff
@@ -1106,7 +1105,6 @@ func (c *Client) CreateOrUpdateClusterRoleBinding(crb *rbacv1.ClusterRoleBinding
 	changed := reflect.DeepEqual(crb.RoleRef, existing.RoleRef)
 	changed = changed || reflect.DeepEqual(crb.Subjects, existing.Subjects)
 	changed = changed || reflect.DeepEqual(crb.Labels, existing.Labels)
-	changed = changed || reflect.DeepEqual(crb.Annotations, existing.Annotations)
```
I believe we have to be a little smarter. If we just ignore the annotations then it is fine, but I think we have to preserve the existing annotations of the object that was found, aka `crb.Annotations = existing.Annotations`.

s-urbaniak replied:
cc @sttts @p0lyn0mial for verification

p0lyn0mial replied:
you could "merge" the annotation map instead of ignoring it.

please have a look at https://github.com/openshift/library-go/blob/master/pkg/operator/resource/resourceapply/rbac.go#L58  this is what we use in our operators.

sttts replied:
Either merging, or restrict deep equal to the annotation "subdomain" you own, e.g. monitoring.openshift.io. But you have allow anything else without wiping it, e.g. `important.customer.com/foo: bar`. It's allowed for every actor to put annotations.

paulfantom replied:
@p0lyn0mial in case of merging annotations, should we follow the same pattern you have of merging whole ObjectMeta? Is there any benefit to such approach?

p0lyn0mial replied:
We don't merge the whole ObjectMeta, just only a few things
```
func EnsureObjectMeta(modified *bool, existing *metav1.ObjectMeta, required metav1.ObjectMeta) {
	SetStringIfSet(modified, &existing.Namespace, required.Namespace)
	SetStringIfSet(modified, &existing.Name, required.Name)
	MergeMap(modified, &existing.Labels, required.Labels)
	MergeMap(modified, &existing.Annotations, required.Annotations)
}
```

paulfantom commented:
Seems like we had an incorrect logic for checking if CRB needs a change. The last commit fixes it by porting [logic used in RB](https://github.com/openshift/cluster-monitoring-operator/blob/master/pkg/client/client.go#L1054-L1059) to do the same

s-urbaniak commented on pkg/client/client.go:
```diff
@@ -1103,12 +1102,9 @@ func (c *Client) CreateOrUpdateClusterRoleBinding(crb *rbacv1.ClusterRoleBinding
-	if !changed {
+	if reflect.DeepEqual(crb.RoleRef, existing.RoleRef) &&
+		reflect.DeepEqual(crb.Subjects, existing.Subjects) &&
+		reflect.DeepEqual(crb.Labels, existing.Labels) {
```
great catch! :tada: :clap:

s-urbaniak replied:
while we're here, do you mind to add some unit testing around this logic?

paulfantom replied:
done

paulfantom commented:
Fixing the annotation issue will be done as part of https://bugzilla.redhat.com/show_bug.cgi?id=1879930

simonpasquier commented on pkg/client/client_test.go:
```diff
@@ -0,0 +1,163 @@
+				t.Fatal(err)
+			}
+
+			err = c.CreateOrUpdateClusterRoleBinding(crb)
```
shouldn't we test that `err` is `nil`?

paulfantom replied:
Yes, I missed that one. Fixed.

simonpasquier commented:
> Fixing CRB re-creation checks by using the same logic as we have for RoleBindings.

Looking at the code it seems that `CreateOrUpdateRoleBinding` should be updated too?

https://github.com/openshift/cluster-monitoring-operator/blob/40bb0e8b5cb588eb27f84e9818be08b0855e15d5/pkg/client/client.go#L1056-L1061

paulfantom commented:
> Looking at the code it seems that CreateOrUpdateRoleBinding should be updated too?

Yes, but we also have another BZ to fix this for all resources: https://bugzilla.redhat.com/show_bug.cgi?id=1879930.

That said I updated `CreateOrUpdateRoleBindings` and rebased.

s-urbaniak commented on pkg/client/client_test.go:
```diff
@@ -0,0 +1,166 @@
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func TestCreateOrUpdateClusterRoleBinding(t *testing.T) {
```
could we add a test that asserts the role binding is not changed if annotations are added?

s-urbaniak replied:
and that annotations are preserved.

paulfantom replied:
added

sttts commented on pkg/client/client.go:
```diff
@@ -1055,11 +1055,15 @@ func (c *Client) CreateOrUpdateRoleBinding(rb *rbacv1.RoleBinding) error {
+	// TODO(paulfantom): currently, we don't set our own annotations for cluster role bindings.
+	// This needs a more elaborated merging logic once we do.
+	// See discussion from https://github.com/openshift/cluster-monitoring-operator/pull/932#discussion_r490089300
+	rb.Annotations = r.Annotations
```
what happens when you set your own annoations? Having a time bomb knows to fail without unit tests or other means to tell is not a good idea.

sttts replied:
isn't annotation merge logic simple enough to just do it here rather than adding two TODOs plus those "means" I mention above? Seems like time well spend to make it correct once and for all.

paulfantom replied:
We are aware and I am working on a fix for this in a follow-up BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1879930. Fix would be applied to all resources, not only CRB.

s-urbaniak replied:
My 2 cents: if the merging logic is not too much effort, I'd pull it in as @sttts suggests. especially given the fact that 4.6 is going to be an LTS release.

@paulfantom do you estimate how much effort it would mean to pull in merging for cluster-role-bindings only? I agree that tackling all the other resources can be done separately as they don't seem to suffer the issue we're observing here.

paulfantom replied:
ok, I added merging Annotations for RBs and CRBs. Other resources will be done in follow-up PR.

s-urbaniak commented on pkg/client/client_test.go:
```diff
@@ -0,0 +1,190 @@
+				t.Fatal(err)
+			}
+
+			updated := reflect.DeepEqual(before, after)
```
nit: the value of `updated` will be true if `before` and `after` do not differ.
however, it will be `false` if `before` and `after` do differ.
shouldn't we call it rather `unchanged`?

paulfantom replied:
makes sense, fixed

sttts commented on pkg/client/client.go:
```diff
@@ -1053,10 +1054,11 @@ func (c *Client) CreateOrUpdateRoleBinding(rb *rbacv1.RoleBinding) error {
-		reflect.DeepEqual(rb.Subjects, r.Subjects) &&
-		reflect.DeepEqual(rb.Labels, r.Labels) &&
-		reflect.DeepEqual(rb.Annotations, r.Annotations) {
+	mergo.Merge(&rb.Annotations, existing.Annotations)
```
this won't remove old annotations. Is that ok?

At least this must be godoc'ed on this func.

sttts replied:
Also this mutates rb.Annotation.

/hold

paulfantom replied:
Yes, this won't remove old annotations. The goal was to merge old/existing ones with what is defined by the operator.

Since `rb` and `crb` is what we hardcode in operator, mutating `rb.Annotation` is how we can update whole object with added Annotations.

I have added a comment to reflect that.

sttts replied:
it is very bad style to have side-effects in such a function, especially when named `CreateOrUpdateRoleBinding`. None of the upstream client funcs does that for a reason. This func should not divert. It is even dangerious if it does because consumers will not expect it and potentially invalidate objects from a cache.

s-urbaniak replied:
I agree here, for the caller this is unexpected. @paulfantom let's make sure we address this in the follow-up. An easy way forward would be to create a deep copy of `rb`.

s-urbaniak commented:
/lgtm once green

s-urbaniak commented:
/lgtm

s-urbaniak commented:
/test e2e-aws-operator
/hold cancel

s-urbaniak commented:
/lgtm

paulfantom commented:
/retest
